datafusion/common/src/cast.rs

```diff
@@ -116,3 +116,36 @@ pub fn as_string_array(array: &dyn Array) -> Result<&StringArray, DataFusionErro
         ))
     })
 }
+
+// Downcast ArrayRef to UInt32Array
+pub fn as_uint32_array(array: &dyn Array) -> Result<&UInt32Array, DataFusionError> {
+    array.as_any().downcast_ref::<UInt32Array>().ok_or_else(|| {
+        DataFusionError::Internal(format!(
+            "Expected a UInt32Array, got: {}",
+            array.data_type()
+        ))
+    })
+}
+
+// Downcast ArrayRef to UInt64Array
+pub fn as_uint64_array(array: &dyn Array) -> Result<&UInt64Array, DataFusionError> {
+    array.as_any().downcast_ref::<UInt64Array>().ok_or_else(|| {
+        DataFusionError::Internal(format!(
+            "Expected a UInt64Array, got: {}",
+            array.data_type()
+        ))
+    })
+}
+
+// Downcast ArrayRef to BooleanArray
+pub fn as_boolean_array(array: &dyn Array) -> Result<&BooleanArray, DataFusionError> {
+    array
+        .as_any()
+        .downcast_ref::<BooleanArray>()
+        .ok_or_else(|| {
+            DataFusionError::Internal(format!(
+                "Expected a BooleanArray, got: {}",
+                array.data_type()
+            ))
+        })
+}
```

Comment:
We have a macro that you can use instead of writing these methods:

```rust
let array = downcast_value!(values, Int32Array)
```

Reply:
maybe we should revert all pr from this issue https://github.com/apache/arrow-datafusion/issues/3152?
cc @alamb @andygrove

Reply:
Well, that is embarrassing 🤦 

Interestingly the `downcast_value` doesn't appear to be used in that many places (around 5 at this time0:
https://github.com/search?q=repo%3Aapache%2Farrow-datafusion+downcast_value&type=code

 I would prefer not to roll back the PRs as they have already simplified the code non trivially.

What is important in my opinion is to use a standard pattern to do this downcasting. I don't have a huge preference between `downcast_value` and `as_boolean_array` , though the `as_boolean_array` might be more discoverable in an IDE that autocompletes

If we are worried about code duplication, perhaps we can use `downcast_value` to implement the `as_boolean_array`, type methods

Reply:
Thanks for the feedback. I just didn't know about this macro. If you feel that this is wrong or not good aesthetically, I will refactor previous PRs with `downcast_value` macro.